FreeRTOS/FreeRTOS-Kernel
# Fix support for stepping tick by xExpectedIdleTime

Description
-----------
This PR fixes support for user implementations of tickless idle that call `vTaskStepTick()` with `xExpectedIdleTime` ticks to step.  The new code ensures `xTickCount` reaches `xNextTaskUnblockTime` inside `xTaskIncrementTick()` instead of inside `vTaskStepTick()`.

Without this PR
--------
**Typical Case: Task wakes up one tick late**
* At `xTickCount` = 10, app task calls `vTaskDelay(5)`
* Idle task calls `vPortSuppressTicksAndSleep(5)` (`xExpectedIdleTime` = 5)
* `vPortSuppressTicksAndSleep()` sleeps 5 ticks and calls `vTaskStepTick(5)` [custom implementation -- not FreeRTOS code]
* [App task remains on delayed list, even though `xNextTaskUnblockTime` == `xTickCount`.]
* One tick period elapses (idle task running the idle loop)
* Tick interrupt occurs, and `xTaskIncrementTick()` sets `xTickCount` to 16 and moves app task to ready list.
* App task wakes up at `xTickCount` = 16 but should have woken at `xTickCount` = 15.

**Rare Case: Assertion failure when xTickCount rolls over**
* Same as example above except starting value for xTickCount is 0xFFFFFFFA instead of 10.
* App task should wake at 0xFFFFFFFF but doesn't (same as above).
* Next tick occurs, and the rollover logic for the tick counter asserts due to the delayed-task list not being empty.

Test Steps
-----------
Choose any low-power demo, and use the `vPortSuppressTicksAndSleep()` version below.  (If there is already a custom version of `vPortSuppressTicksAndSleep()` in the demo, remove it.)  Strip the demo down to just one task, with no active timers, then insert a call to function `test()` from that task.  Without this PR, the `configASSERT()` statement in `test()` fails.  Then remove that statement to see the next failure at `vTaskDelay( xDelta );` -- another assertion failure.  After the PR, there are no assertion failures.  Note the actual demo won't function normally with this test code even after the PR because this demo version of  `vPortSuppressTicksAndSleep()` doesn't actually sleep.  It just illustrates the API issue.
 
```
void vPortSuppressTicksAndSleep( TickType_t xExpectedIdleTime )
{
    /* This version of vPortSuppressTicksAndSleep() exercises the tickless API
     * only.  The specific case being simulated is a sleep that completes fully
     * with no early wake-up.  The sleep ends via timer other than the tick
     * timer, so no tick is pending. */
    vTaskStepTick( xExpectedIdleTime );
}

void test()
{
    /* Get to the beginning of a tick period. */
    vTaskDelay( 1 );

    /* Sleep (tickless idle) for ten ticks. */
    TickType_t xTickCount = xTaskGetTickCount();
    vTaskDelay( 10 );

    /* Make sure we slept only ten ticks. */    
    TickType_t xDelta = xTaskGetTickCount() - xTickCount;
    configASSERT( xDelta == 10 );  /* fails without PR 73 */

    /* Determine how many ticks until just before the tick rolls over. */
    xTickCount += xDelta;
    xDelta = portMAX_DELAY - xTickCount;

    /* Sleep (tickless idle) until just before the tick rolls over. */
    vTaskDelay( xDelta );  /* causes assert in tasks.c without PR 73 */
}
```

Related Issue
-----------
https://forums.freertos.org/t/assert-in-tasks-c-2611-on-xtickcount-wrap-around/9984

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
